drivers/sensor/sensor_sim/sensor_sim.c

```diff
@@ -6,27 +6,106 @@
 
 #include <drivers/gpio.h>
 #include <init.h>
-#include <drivers/sensor.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <logging/log.h>
 
-#include "sensor_sim.h"
+#include "sensor_sim_priv.h"
+#include <drivers/sensor_sim.h>
 
-#include <math.h>
-#ifndef M_PI
-  #define M_PI 3.14159265358979323846
-#endif
+#define ACCEL_CHAN_COUNT	3
 
 LOG_MODULE_REGISTER(sensor_sim, CONFIG_SENSOR_SIM_LOG_LEVEL);
 
-static const double base_accel_samples[3] = {0.0, 0.0, 0.0};
-static double accel_samples[3];
+#define ACCEL_DEFAULT_TYPE		WAVE_GEN_TYPE_SINE
+#define ACCEL_DEFAULT_AMPLITUDE		20.0
+#define ACCEL_DEFAULT_PERIOD_MS		10000
+
+static struct wave_gen_param accel_param[ACCEL_CHAN_COUNT];
+struct k_mutex accel_param_mutex;
+
+static double accel_samples[ACCEL_CHAN_COUNT];
 
 static double temp_sample;
 static double humidity_sample;
 static double pressure_sample;
 
+/**
+ * @typedef generator_function
+ * @brief Function used to generate sensor value for given channel.
+ *
+ * @param chan[in]	Selected sensor channel.
+ * @param out_val[out]	Pointer to the variable that is used to store result.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+typedef int (*generator_function)(enum sensor_channel chan, double *out_val);
+
+/*
+ * @brief Function used to get wave parameters for given sensor channel.
+ *
+ * @param chan[in]	Selected sensor channel.
+ *
+ * @return Pointer to the structure describing parameters of generated wave.
+ */
+static struct wave_gen_param *get_wave_params(enum sensor_channel chan)
+{
+	struct wave_gen_param *dest = NULL;
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		dest = &accel_param[0];
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		dest = &accel_param[1];
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		dest = &accel_param[2];
+		break;
+	default:
+		break;
+	}
+
+	return dest;
+}
+
+int sensor_sim_set_wave_param(enum sensor_channel chan, const struct wave_gen_param *set_params)
+{
+	if (!IS_ENABLED(CONFIG_SENSOR_SIM_ACCEL_WAVE)) {
+		return -ENOTSUP;
+	}
+
+	struct wave_gen_param *dest = get_wave_params(chan);
+
+	if (!dest) {
+		return -ENOTSUP;
+	}
+
+	if ((set_params->period_ms == 0) || (set_params->type >= WAVE_GEN_TYPE_COUNT)) {
+		return -EINVAL;
+	}
+
+	int err = k_mutex_lock(&accel_param_mutex, K_FOREVER);
```

Comment:
you don't need to assert that as `CONFIG_ASSERT_ON_ERRORS` will do it globally

There is also little point; this is sensor_sim code, rather non-critical element of the system whereas neither in Zephyr's subsys nor drivers nobody checks this return value.

Reply:
the same goes to unlock

Reply:
Removed error check